File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    
    double fastPow(double x, long long N){
        if(N == 0) return 1.0;
        
        double half = fastPow(x, N/2);
        
        if(N % 2 == 0){
            return half * half;
        }else{
            return half*half*x;
        }
    }
    double myPow(double x, int n) {
        long long N = n;
        
        if(N<0){
            N = -N;
            x = 1/x;
        }
        
//         double ans = 1.0;
        
//         for(long long i=1; i<=N; i++){
//             ans *= x;
//         }
        
        return fastPow(x, N);
    }
};

File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, int n) {
        if(n == Integer.MIN_VALUE){
            x = 1/x;
            n = Integer.MAX_VALUE;
            return myPow(x, n) * x;
        }

        else if(n < 0){
            x = 1/x;
            n = -n;
            return myPow(x, n);
        }

        else if(n == 0) {
            return 1f;
        }

        double res = 1f;

        while(n>0){
            if(n%2==1){
                res*=x;
            }
            n/=2;
            x*=x;
        }

        return res;
    }
}